### ISSUE TITLE
expose_inputs tries to use deepcopy on stored data

### ISSUE THREAD

**Issue by DropD** [aiidateam/aiida_core#1741]
## How to reproduce
* aiida-core: current develop (after commit 777c16)

Run a workchain which exposes inputs of a lower level one. Example: 'vasp.relax' from aiida-vasp.

## Observed behaviour
```
.tox/aiida_dev/lib/python2.7/site-packages/aiida/work/launch.py:43: in run
    return runner.run(process, *args, **inputs)
.tox/aiida_dev/lib/python2.7/site-packages/aiida/work/runners.py:247: in run
    result, _ = self._run(process, *args, **inputs)
.tox/aiida_dev/lib/python2.7/site-packages/aiida/work/runners.py:234: in _run
    process = self.instantiate_process(process, *args, **inputs)
.tox/aiida_dev/lib/python2.7/site-packages/aiida/work/runners.py:191: in instantiate_process
    process = process_class(runner=self, inputs=inputs)
.tox/aiida_dev/lib/python2.7/site-packages/plumpy/base/state_machine.py:184: in __call__
    inst = super(StateMachineMeta, cls).__call__(*args, **kwargs)
.tox/aiida_dev/lib/python2.7/site-packages/aiida/work/workchain.py:54: in __init__
    inputs=inputs, logger=logger, runner=runner, enable_persistence=enable_persistence)
.tox/aiida_dev/lib/python2.7/site-packages/aiida/work/processes.py:88: in __init__
    inputs=self.spec().inputs.serialize(inputs),
.tox/aiida_dev/lib/python2.7/site-packages/plumpy/processes.py:161: in spec
    cls.define(cls._spec)
.tox/aiida_dev/lib/python2.7/site-packages/aiida_vasp/workflows/relax.py:97: in define
    spec.expose_inputs(VaspBaseWf, namespace='restart', include=['clean_workdir'])
.tox/aiida_dev/lib/python2.7/site-packages/plumpy/process_spec.py:227: in expose_inputs
    namespace_options=namespace_options,
.tox/aiida_dev/lib/python2.7/site-packages/plumpy/process_spec.py:292: in _expose_ports
    namespace_options)
.tox/aiida_dev/lib/python2.7/site-packages/plumpy/ports.py:419: in absorb
    self[port_name] = copy.deepcopy(port)
../../miniconda2/envs/aiida-dev/lib/python2.7/copy.py:190: in deepcopy
    y = _reconstruct(x, rv, 1, memo)
../../miniconda2/envs/aiida-dev/lib/python2.7/copy.py:334: in _reconstruct
    state = deepcopy(state, memo)
../../miniconda2/envs/aiida-dev/lib/python2.7/copy.py:163: in deepcopy
    y = copier(x, memo)
../../miniconda2/envs/aiida-dev/lib/python2.7/copy.py:257: in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
../../miniconda2/envs/aiida-dev/lib/python2.7/copy.py:174: in deepcopy
    y = copier(memo)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <[AttributeError("DbAttribute with key value for node 500 not found in db") raised in repr()] SafeRepr object at 0x7f91e3fa2200>
memo = {72535040: <class 'aiida.orm.data.bool.Bool'>, 74815232: <class 'aiida.work.ports.InputPort'>, 140264157025488: <aiida.work.ports.InputPort object at 0x7f91cc3ab210>, 140264157653776: (<class 'aiida.work.ports.InputPort'>,), ...}

    def __deepcopy__(self, memo):
        """
            Create a clone of the Data node by pipiong through to the clone method and return the result.
    
            :returns: an unstored clone of this Data node
            """
        if self.is_stored:
>           raise NotImplementedError('deep copying a stored Data node is not supported, use Data.clone() instead')
E           NotImplementedError: deep copying a stored Data node is not supported, use Data.clone() instead
```

## Expected behaviour
The machinery for expose_inputs takes into account the latest changes and does not raise an error.

**Comment by muhrin** [comment]
I've had a look back at this issues and had a chat with @sphuber .

This arises, as noted above, because `plumpy` is trying to deepcopy the input ports and crucially their default values into the child that is exposing.  This is a problem if the default value is already stored because in this case deepcopying is disabled given that's it's a little confusing exactly what this should do.   We provide a `clone` method that makes a full copy of the stored node with a new `uuid` and in an unstored state.

What I will do is just allow deepcopying and call through to `clone` for now.

What I would like to do: make all default immutable and keep the ban on deepcopying stored nodes.  Then we could just use the same reference in exposed inputs and have no copy what so ever given that the value is assumed to be immutable.  This would also be closer to what python does with default values.  However, for someone writing a workflow there is no way to provide a stored input as a default without, say, storing it at the module level which would incur a new node each time the module is loaded.  Not cool.

**Comment by sphuber** [comment]
Fixed in PR #2347
